getredash/redash
# Add generic OIDC authentication flow

## What type of PR is this? 


- [ ] New Alert Destination

## Description

### Generic OIDC login

_<ins>Note: Seeking feedback</ins>_

### Problem

Currently, Redash is limited to supporting authentication via Google's OIDC. Expanding the authentication support to a generic OIDC flow increases Redash's accessibility and flexibility, making it a more inclusive tool for diverse environments.

### Implementation Steps

To fully implement and integrate the generic OIDC authentication flow into Redash, the following steps need to be taken (Open for feedback):

- [x] **Make OIDC Discovery Endpoint Configurable**: This flexibility allows Redash to dynamically adapt to different IdPs by simply modifying configuration parameters.

- [x] **Implement Routes for OIDC Authentication**: Develop new routes within the application to handle the authentication process via OIDC.

- [x] **Modify Login Templates**: Update the UI components, specifically the login page templates, to include an OIDC login button.

- [ ] **Manual Testing with Multiple OIDC IdPs**: To ensure compatibility and robustness, manually test the new authentication flow with multiple OIDC IdPs, including but not limited to Google, AWS, Okta, and Auth0. (Tested with Google and AWS Cognito)

- [x] **Implement Unit Tests**

- [ ] **Write Documentation for OIDC Configuration**: Create detailed documentation to assist users in configuring OIDC with Redash. This documentation should cover configuration steps for different IdPs.

- [x] **Mutate OIDC to include domain check**: (not sure about this one) Current Google OIDC flow has domain verification, see if this can be made generic.

- [ ] **Replace Google OIDC with generic one**

## How is this tested?



For testing manually, following steps should be more are less common among different IdPs.

1. Configure an IdP with OIDC federation protocol and get the discovery endpoint, client id and secret id.
2. Set 3 environment variables accordingly:
 - REDASH_OIDC_CLIENT_ID
 - REDASH_OIDC_CLIENT_SECRET
 - REDASH_OIDC_ISSUER_URL
 3. Check if Login with OIDC button available in the login page.
 4. Login
 
 - [x] AWS Cognito
 - [x] Google
 - [x] Okta
 - [ ] Auth0
 - [ ] Azure


## Related Tickets & Documents

#6781 

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
<img width="524" alt="Screenshot 2024-02-27 at 2 43 36 PM" src="https://github.com/getredash/redash/assets/10260127/89036a80-231b-4923-9152-e08d7f3aaaff">
<img width="969" alt="Screenshot 2025-02-18 at 11 40 13 PM" src="https://github.com/user-attachments/assets/2dafbc7a-a6ce-42d8-9f8c-61627716ed9e" />
